File: booking/service.py
"""Бизнес-логика бронирования кабинетов.

Модуль не знает ни о CLI, ни о деталях SQL: он оперирует моделями и
возбуждает доменные исключения из :mod:`booking.errors`.
"""

from __future__ import annotations

import sqlite3
from dataclasses import dataclass
from datetime import datetime
from typing import List, Optional

from booking import repository
from booking.config import AppConfig
from booking.db import transaction
from booking.errors import (
    BookingNotFoundError,
    RoomBusyError,
    RoomNotFoundError,
    ValidationError,
)
from booking.models import Booking, Room
from booking.notifier import build_message, get_notifier
from booking.slots import DB_DATE_FORMAT, TimeSlot, validate_email

#: Максимальная длительность одной брони.
MAX_DURATION_MINUTES = 12 * 60


@dataclass(frozen=True)
class RoomAvailability:
    """Результат проверки одного кабинета на заданный интервал."""

    room: Room
    conflicts: List[Booking]

    @property
    def is_free(self) -> bool:
        """Признак того, что кабинет свободен."""
        return not self.conflicts

    @property
    def busy_until(self) -> Optional[datetime]:
        """Момент освобождения кабинета внутри запрошенного интервала."""
        if not self.conflicts:
            return None
        return max(item.slot.end for item in self.conflicts)


class BookingService:
    """Операции над кабинетами и бронями."""

    def __init__(
        self, conn: sqlite3.Connection, config: AppConfig
    ) -> None:
        """Сохранить подключение к БД и конфигурацию."""
        self._conn = conn
        self._config = config

    # ------------------------------------------------------------------
    # Справочники
    # ------------------------------------------------------------------

    def rooms(self) -> List[Room]:
        """Вернуть список кабинетов офиса."""
        return repository.list_rooms(self._conn)

    def get_room(self, room_id: int) -> Room:
        """Вернуть кабинет или возбудить :class:`RoomNotFoundError`."""
        room = repository.get_room(self._conn, room_id)
        if room is None:
            raise RoomNotFoundError(room_id)
        return room

    # ------------------------------------------------------------------
    # Проверка занятости
    # ------------------------------------------------------------------

    def check(
        self, slot: TimeSlot, room_id: Optional[int] = None
    ) -> List[RoomAvailability]:
        """Проверить занятость одного или всех кабинетов на интервал.

        Если ``room_id`` не задан, проверяются все кабинеты офиса.
        """
        self._validate_slot(slot)
        rooms = [self.get_room(room_id)] if room_id else self.rooms()
        return [
            RoomAvailability(
                room=room,
                conflicts=repository.find_conflicts(
                    self._conn, room.id, slot
                ),
            )
            for room in rooms
        ]

    def is_free(self, room_id: int, slot: TimeSlot) -> bool:
        """Короткая проверка: свободен ли конкретный кабинет."""
        return self.check(slot, room_id)[0].is_free

    # ------------------------------------------------------------------
    # Бронирование
    # ------------------------------------------------------------------

    def book(
        self,
        room_id: int,
        person_name: str,
        person_email: str,
        slot: TimeSlot,
        purpose: str = "",
    ) -> Booking:
        """Забронировать кабинет на интервал.

        Проверка занятости и вставка выполняются в одной транзакции
        ``BEGIN IMMEDIATE``, поэтому два параллельных процесса не смогут
        занять один и тот же интервал.

        :raises RoomBusyError: кабинет уже занят на это время.
        """
        self._validate_slot(slot)
        name = person_name.strip()
        if not name:
            raise ValidationError("Имя сотрудника не может быть пустым.")
        email = validate_email(person_email)
        room = self.get_room(room_id)

        with transaction(self._conn):
            conflicts = repository.find_conflicts(self._conn, room.id, slot)
            if conflicts:
                raise RoomBusyError(room, conflicts)
            booking_id = repository.insert_booking(
                self._conn, room.id, name, email, slot, purpose.strip()
            )
        booking = repository.get_booking(self._conn, booking_id)
        assert booking is not None  # запись только что вставлена
        return booking

    def cancel(self, booking_id: int) -> Booking:
        """Отменить бронь и вернуть удалённую запись."""
        booking = self.get_booking(booking_id)
        with transaction(self._conn):
            repository.delete_booking(self._conn, booking_id)
        return booking

    def get_booking(self, booking_id: int) -> Booking:
        """Вернуть бронь или возбудить :class:`BookingNotFoundError`."""
        booking = repository.get_booking(self._conn, booking_id)
        if booking is None:
            raise BookingNotFoundError(booking_id)
        return booking

    def agenda(
        self,
        room_id: Optional[int] = None,
        on_date: Optional[datetime] = None,
        include_past: bool = False,
    ) -> List[Booking]:
        """Вернуть расписание броней с фильтрами по кабинету и дате."""
        if room_id is not None:
            self.get_room(room_id)
        date_filter = (
            on_date.strftime(DB_DATE_FORMAT) if on_date else None
        )
        return repository.list_bookings(
            self._conn, room_id, date_filter, include_past
        )

    # ------------------------------------------------------------------
    # Уведомления
    # ------------------------------------------------------------------

    def notify(self, booking_id: int, dry_run: bool = False) -> str:
        """Отправить письмо владельцу брони.

        В письме указаны дата, время и номер кабинета. При успешной
        отправке проставляется отметка ``notified_at``.
        """
        booking = self.get_booking(booking_id)
        room = self.get_room(booking.room_id)
        message = build_message(booking, room, self._config.smtp)
        notifier = get_notifier(self._config.smtp, dry_run)
        result = notifier.send(message)
        if not dry_run:
            with transaction(self._conn):
                repository.mark_notified(self._conn, booking_id)
        return result

    # ------------------------------------------------------------------
    # Внутренние проверки
    # ------------------------------------------------------------------

    def _validate_slot(self, slot: TimeSlot) -> None:
        """Проверить интервал на длительность и рабочие часы."""
        if slot.duration_minutes > MAX_DURATION_MINUTES:
            raise ValidationError(
                "Бронь не может длиться дольше "
                f"{MAX_DURATION_MINUTES // 60} часов."
            )
        start_ok = slot.start.time() >= self._config.workday_start
        end_ok = (
            slot.end.time() <= self._config.workday_end
            or slot.end.time() == datetime.min.time()
        )
        if not (start_ok and end_ok):
            raise ValidationError(
                "Время вне рабочих часов офиса "
                f"({self._config.workday_start:%H:%M}-"
                f"{self._config.workday_end:%H:%M})."
            )

File: internal/booking/service.go
package booking

import (
	"fmt"
	"io"
	"strings"
)

// maxDurationMinutes — максимальная длительность одной брони.
const maxDurationMinutes = 12 * 60

// Service содержит бизнес-логику бронирования. Он не знает ни о выводе
// в консоль, ни о деталях SQL.
type Service struct {
	store  *Store
	config Config
	// notifier подменяет отправителя писем; если nil, он выбирается
	// по конфигурации.
	notifier Notifier
	// out — поток, куда печатается письмо, когда SMTP не настроен.
	out io.Writer
}

// NewService создаёт сервис поверх хранилища.
func NewService(store *Store, config Config) *Service {
	return &Service{store: store, config: config}
}

// SetNotifier подменяет отправителя уведомлений.
func (s *Service) SetNotifier(n Notifier) { s.notifier = n }

// SetOutput задаёт поток для вывода письма при ненастроенном SMTP.
func (s *Service) SetOutput(w io.Writer) { s.out = w }

// Rooms возвращает список кабинетов офиса.
func (s *Service) Rooms() ([]Room, error) { return s.store.Rooms() }

// Room возвращает кабинет по номеру.
func (s *Service) Room(id int) (Room, error) { return s.store.Room(id) }

// Booking возвращает бронь по номеру.
func (s *Service) Booking(id int) (Booking, error) { return s.store.Booking(id) }

// Check проверяет занятость кабинетов на интервал.
//
// Если roomID равен нулю, проверяются все кабинеты офиса.
func (s *Service) Check(slot TimeSlot, roomID int) ([]Availability, error) {
	if err := s.validateSlot(slot); err != nil {
		return nil, err
	}
	var rooms []Room
	if roomID > 0 {
		room, err := s.store.Room(roomID)
		if err != nil {
			return nil, err
		}
		rooms = []Room{room}
	} else {
		var err error
		if rooms, err = s.store.Rooms(); err != nil {
			return nil, err
		}
	}

	results := make([]Availability, 0, len(rooms))
	for _, room := range rooms {
		conflicts, err := s.store.Conflicts(room.ID, slot)
		if err != nil {
			return nil, err
		}
		results = append(results, Availability{Room: room, Conflicts: conflicts})
	}
	return results, nil
}

// IsFree сообщает, свободен ли конкретный кабинет на интервал.
func (s *Service) IsFree(roomID int, slot TimeSlot) (bool, error) {
	results, err := s.Check(slot, roomID)
	if err != nil {
		return false, err
	}
	return results[0].IsFree(), nil
}

// Book бронирует кабинет на интервал.
//
// Если кабинет занят, возвращается *RoomBusyError со списком
// пересекающихся броней.
func (s *Service) Book(roomID int, name, email string, slot TimeSlot,
	purpose string) (Booking, error) {

	if err := s.validateSlot(slot); err != nil {
		return Booking{}, err
	}
	name = strings.TrimSpace(name)
	if name == "" {
		return Booking{}, validationErrorf("имя сотрудника не может быть пустым")
	}
	email, err := ValidateEmail(email)
	if err != nil {
		return Booking{}, err
	}
	room, err := s.store.Room(roomID)
	if err != nil {
		return Booking{}, err
	}

	id, err := s.store.InsertBooking(room, name, email, slot,
		strings.TrimSpace(purpose))
	if err != nil {
		return Booking{}, err
	}
	return s.store.Booking(id)
}

// Cancel отменяет бронь и возвращает удалённую запись.
func (s *Service) Cancel(id int) (Booking, error) {
	booking, err := s.store.Booking(id)
	if err != nil {
		return Booking{}, err
	}
	if err := s.store.DeleteBooking(id); err != nil {
		return Booking{}, err
	}
	return booking, nil
}

// Agenda возвращает расписание броней с фильтрами.
func (s *Service) Agenda(roomID int, onDate string,
	includePast bool) ([]Booking, error) {

	if roomID > 0 {
		if _, err := s.store.Room(roomID); err != nil {
			return nil, err
		}
	}
	return s.store.ListBookings(roomID, onDate, includePast)
}

// Notify отправляет владельцу брони письмо с датой, временем и номером
// кабинета. При успешной отправке проставляется отметка notified_at.
func (s *Service) Notify(bookingID int, dryRun bool) (string, error) {
	booking, err := s.store.Booking(bookingID)
	if err != nil {
		return "", err
	}
	room, err := s.store.Room(booking.RoomID)
	if err != nil {
		return "", err
	}

	notifier := s.notifier
	if notifier == nil {
		notifier = NewNotifier(s.config.SMTP, dryRun, s.out)
	}
	message := BuildMessage(booking, room, s.config.SMTP)
	result, err := notifier.Send(message)
	if err != nil {
		return "", err
	}
	if !dryRun {
		if err := s.store.MarkNotified(bookingID); err != nil {
			return "", err
		}
	}
	return result, nil
}

// validateSlot проверяет интервал на длительность и рабочие часы.
func (s *Service) validateSlot(slot TimeSlot) error {
	if slot.Minutes() > maxDurationMinutes {
		return validationErrorf("бронь не может длиться дольше %d часов",
			maxDurationMinutes/60)
	}
	startHour, startMin, err := ParseClock(s.config.DayStart)
	if err != nil {
		return err
	}
	endHour, endMin, err := ParseClock(s.config.DayEnd)
	if err != nil {
		return err
	}

	slotStart := slot.Start.Hour()*60 + slot.Start.Minute()
	slotEnd := slot.End.Hour()*60 + slot.End.Minute()
	if slotEnd == 0 {
		slotEnd = 24 * 60
	}
	dayStart := startHour*60 + startMin
	dayEnd := endHour*60 + endMin

	if slotStart < dayStart || slotEnd > dayEnd {
		return validationErrorf("время вне рабочих часов офиса (%s-%s)",
			s.config.DayStart, s.config.DayEnd)
	}
	return nil
}

// describeBooking формирует подробное описание брони для вывода.
func describeBooking(b Booking, room Room) string {
	var sb strings.Builder
	fmt.Fprintf(&sb, "  Бронь №%d\n", b.ID)
	fmt.Fprintf(&sb, "  Кабинет:   №%d — %s\n", room.ID, room.Name)
	fmt.Fprintf(&sb, "  Дата:      %s\n", b.Slot.Start.Format("02.01.2006"))
	fmt.Fprintf(&sb, "  Время:     %s (%d мин.)\n", b.Slot.HumanTime(),
		b.Slot.Minutes())
	fmt.Fprintf(&sb, "  Сотрудник: %s", b.Person())
	if b.Purpose != "" {
		fmt.Fprintf(&sb, "\n  Тема:      %s", b.Purpose)
	}
	return sb.String()
}
